### PR TITLE
[#3] Add `returnTicker` API

### PR DESCRIPTION
**Pull request by freeatnet**
See #3 and previous work in #18.

With massive shouts to @JonathonDunford for previous digging.

### CODE REVIEW

**Comment by JonathonDunford** [comment]
return of the ticker LOL

File: alembic/versions/e032ff832b82_create_ticker_table.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+
+# revision identifiers, used by Alembic.
+revision = 'e032ff832b82'
+down_revision = '7719b4610d7e'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
```
Comment (JonathonDunford) [inline alembic/versions/e032ff832b82_create_ticker_table.py:23]: Looks beautiful. Like something I would write :P

File: app/services/ticker.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+from ..app import App
+from ..src.erc20_token import ERC20Token
+
+logger = logging.getLogger('services.ticker')
+logger.setLevel(logging.DEBUG)
+
+getcontext().prec = 8
+
+# TODO: finally extract these into constants
```
Comment (JonathonDunford) [inline app/services/ticker.py:16]: One of these days, we will make this a constant

File: app/services/ticker.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+ZERO_ADDR = "0x0000000000000000000000000000000000000000"
+ZERO_ADDR_BYTES = Web3.toBytes(hexstr=ZERO_ADDR)
+
+tokens_queue = Queue()
+# TODO: Populate from our own DB
```
Comment (JonathonDunford) [inline app/services/ticker.py:21]: Yeah, this is pretty important. Is there any way to import it from the main config for now?

File: app/services/ticker.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    async with App().db.acquire_connection() as conn:
+        return await conn.fetchrow("""
+            SELECT
+                COALESCE(SUM(CASE WHEN "token_get" = $1 THEN "amount_give" ELSE "amount_get" END), 0) AS quote_volume,
+                COALESCE(SUM(CASE WHEN "token_get" = $1 THEN "amount_get" ELSE "amount_give" END), 0) AS base_volume
+            FROM trades
+            WHERE ("token_give" = $2 OR "token_get" = $2)
+                AND "date" >= NOW() - '1 day'::INTERVAL
+            """,
+            Web3.toBytes(hexstr=ZERO_ADDR), # TODO: We can probably do with just token address
```
Comment (JonathonDunford) [inline app/services/ticker.py:44]: This is future proof allowing a secondary address to check against though.

File: app/services/ticker.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            """, Web3.toBytes(hexstr=token_hexstr))
+
+async def get_market_spread(token_hexstr, current_block):
+    """
+    Given a token address, returns the lowest ask and the highest bid.
+    """
+
+    async with App().db.acquire_connection() as conn:
+        return await conn.fetchrow("""
+            SELECT
```
Comment (JonathonDunford) [inline app/services/ticker.py:69]: Beautiful. Has this SQL been tested?

File: app/services/ticker.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            ticker_info["last"] = price
+    else: # No last price available
+        ticker_info["last"] = None
+
+    # TODO: percent_change
+
+    spread = await get_market_spread(token_addr, current_block)
+    if spread["ask"]:
+        # Spread comes back in token/base native values, denormalize it here
+        ticker_info["ask"] = spread["ask"] * base_contract.denormalize_value(1.0) \
```
Comment (JonathonDunford) [inline app/services/ticker.py:159]: Why are we multiplying by denormalized 1?

File: app/services/ticker.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    logger.debug(ticker_info)
+    await save_ticker(ticker_info)
+
+
+async def main():
+    while True:
+        token = tokens_queue.get()
+        await update_ticker(token)
+        tokens_queue.put(token)
+        await asyncio.sleep(2.0)
```
Comment (JonathonDunford) [inline app/services/ticker.py:179]: How did you arrive at 2 seconds?

File: app/services/websocket_server.py [hunk 8]
```diff
@@ -269,13 +270,48 @@ def format_order(record):
...
+async def get_tickers():
+    async with App().db.acquire_connection() as conn:
+        return await conn.fetch("""
+            SELECT *
+            FROM tickers
+            ORDER BY token_address
+            """)
+
+def ticker_key(ticker):
+    return "{}_{}".format("ETH", Web3.toHex(ticker["token_address"])[:9])
```
Comment (JonathonDunford) [inline app/services/websocket_server.py]: Maybe a little comment saying what's going on here.

File: app/services/websocket_server.py [hunk 9]
```diff
@@ -269,13 +270,48 @@ def format_order(record):
...
+        baseVolume=str(ticker["base_volume"]).lower(),
+        last=str(ticker["last"]).lower() if ticker["last"] else None,
+        bid=str(ticker["bid"]).lower() if ticker["bid"] else None,
+        ask=str(ticker["ask"]).lower() if ticker["ask"] else None,
+        updated=ticker["updated"].isoformat()
+    )
+
+def format_tickers(tickers):
+    return {
+        ticker_key(ticker): format_ticker(ticker)
```
Comment (JonathonDunford) [inline app/services/websocket_server.py:303]: Python voodoo magic

File: app/services/websocket_server.py [hunk 10]
```diff
@@ -269,13 +270,48 @@ def format_order(record):
...
+        for ticker in tickers
+    }
+
+@routes.get('/returnTicker')
```
Comment (JonathonDunford) [inline app/services/websocket_server.py:307]: OOOH Http too?

**Review by JonathonDunford — APPROVED** [review]
Most comments are for clarification. Nothing holding us back from pushing.

File: app/services/ticker.py (hunk 5 above)
Comment (freeatnet) [reply app/services/ticker.py:69]: Yeap, I've snuck it into production at this point, even.

File: app/services/ticker.py (hunk 4 above)
Comment (freeatnet) [reply app/services/ticker.py:44]: Fair enough!

File: app/services/ticker.py (hunk 3 above)
Comment (freeatnet) [reply app/services/ticker.py:21]: Not really.

File: app/services/ticker.py (hunk 7 above)
Comment (freeatnet) [reply app/services/ticker.py:179]: "Update every ticker about every half hour", with N_tokens = 1000 gives us ~2 seconds, ignoring processing time. Open for updates and discussion though.

File: app/services/websocket_server.py (hunk 10 above)
Comment (freeatnet) [reply app/services/websocket_server.py:307]: Indeed!

File: app/services/websocket_server.py (hunk 9 above)
Comment (freeatnet) [reply app/services/websocket_server.py:303]: [dictionary comprehensions](http://www.diveintopython3.net/comprehensions.html#dictionarycomprehension) 😄

File: app/services/ticker.py (hunk 7 above)
Comment (JonathonDunford) [reply app/services/ticker.py:179]: Can we run multiple threads of this?

File: app/services/websocket_server.py (hunk 9 above)
Comment (JonathonDunford) [reply app/services/websocket_server.py:303]: AKA magic

File: app/services/ticker.py (hunk 6 above)
Comment (freeatnet) [reply app/services/ticker.py:159]: Normally, we'd convert `amount_get` and `amount_give` and then compute the price; however, `ask` and `bid` returned by the DB are max/min prices for their respective sides of the orderbook, so we convert them to eth-values using this formula.

File: app/services/ticker.py (hunk 7 above)
Comment (freeatnet) [reply app/services/ticker.py:179]: We could. Do we need to?

**Pull request merged**